### PR TITLE
Deploy simulation and gateway servers as part of test pipeline

### PR DESCRIPTION
**Pull request by dan-silver**
This PR deploys a linux gateway vm and a simulation server to the test pipeline using the existing ```azuredeploy.simulation.json``` ARM template. This is required since many of the test cases we're adding later require a gateway device to be connected.

Since IAI is deployed in the test pipeline and deploy.ps1 isn't used, two resources are not created that are required for ```azuredeploy.simulation.json``` to work - DPS and an MSI. Right now this works and I've gotten past a few issues but it's not a good solution. I've extracted the DPS and MSI creation into individual ARM templates that are deployed before the simulation servers. Feedback on this approach is welcome. If this is useful we could refactor ```azuredeploy.standard.json``` to use the new DPS template instead of having two copies.

### CODE REVIEW

File: tools/e2etesting/DeploySimulation.ps1 [hunk 1]
```diff
@@ -0,0 +1,71 @@
+Param(
+    $keyVaultName,
+    $resourceGroupName,
+    $branchName
+)
+
+$branchName = $branchName.Replace('refs/heads/', '')
+$templateDir = [System.IO.Path]::Combine($PSScriptRoot, "../../deploy/templates")
```
Comment (dan-silver) [inline tools/e2etesting/DeploySimulation.ps1:8]: I'm not a fan of these hard coded paths. Will try to find a better way

**Review by markusstuhler — APPROVED** [review]

**Comment by mregen** [comment]
>       azureSubscription: 'IOT-OPC-WALLS-MSI'

could you make this a parameter? I see it multiple times, and maybe it should be a variable in the devops pipeline or of the local dev.

---
Refers to: tools/e2etesting/pipeline.yml:39 in 8024709. [](commit_id = 8024709efd5acd9369be4c0d5fc87fad7307554c, deletion_comment = False)

**Review by mregen — APPROVED** [review]
:shipit:

File: tools/e2etesting/DeploySimulation.ps1 [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+if ($msiDeployment.ProvisioningState -ne "Succeeded") {
+    Write-Error "Deployment $($msiDeployment.ProvisioningState)." -ErrorAction Stop
+}
+
+Write-Host "Created MSI $($msiDeployment.Parameters.managedIdentityName.Value) with resource id $($msiDeployment.Outputs.managedIdentityResourceId.Value)"
+
+# Configure the keyvault
+# Allow the MSI to access keyvault
+# https://github.com/Azure/azure-powershell/issues/10029 for why -BypassObjectIdValidation is needed
+Set-AzKeyVaultAccessPolicy -VaultName $keyVaultName -ObjectId $msiDeployment.Outputs.managedIdentityPrincipalId.Value -PermissionsToSecrets get,list,set,delete -PermissionsToKeys get,list,sign,unwrapKey,wrapKey,create -PermissionsToCertificates get,list,update,create,import -BypassObjectIdValidation
```
[the remark this answers is not in the document]
Comment (dan-silver) [reply tools/e2etesting/DeploySimulation.ps1:47]: Nope, I'll figure out the minimum set. Good idea!

**Review by jyotig-ms — APPROVED** [review]

**Pull request merged**